src/fiber/graph.rs

```diff
@@ -1005,10 +1030,70 @@ where
         if result.is_empty() || current != target {
             return Err(PathFindError::PathFind("no path found".to_string()));
         }
+        if let Some((node, channel)) = last_edge {
+            result.push(PathEdge {
+                target: node,
+                channel_outpoint: channel.clone(),
+            })
+        }
 
         Ok(result)
     }
 
+    fn adjust_target_for_route_self(
+        &self,
+        hop_hint_map: &HashMap<(Pubkey, bool), OutPoint>,
+        amount: u128,
+        source: Pubkey,
+        target: Pubkey,
+    ) -> Result<(Pubkey, u64, Option<(Pubkey, OutPoint)>), PathFindError> {
+        let direct_channels: Vec<(Pubkey, Pubkey, &ChannelInfo, &ChannelUpdateInfo)> = self
+            .get_node_inbounds(source)
+            .filter(|(_, _, channel_info, _)| {
+                if let Some(channel) = hop_hint_map.get(&(source, true)) {
+                    // if there is a hop hint for node -> source,
+                    // try to use this channel as the last candidate hop for route self
+                    // and we event don't check the direct balance of channel,
+                    // hop hint's priority is higher than direct balance
+                    return channel == channel_info.out_point();
+                }
+                if let Some(channel) = hop_hint_map.get(&(source, false)) {
+                    // if there is a hop hint for source -> node,
+                    // then we can not set this node as the last candidate hop for route self
+                    // so skip this channel
+                    if channel == channel_info.out_point() {
+                        return false;
+                    }
+                }
+                if let Some(state) = self
+                    .store
+                    .get_channel_state_by_outpoint(&channel_info.out_point())
+                {
+                    let balance = state.to_remote_amount;
+                    return balance >= amount;
+                }
+                // normal code path will not reach here, we must can get balance for direct channels
+                // anyway, check the capacity here for safety
+                return channel_info.capacity() >= amount;
+            })
+            .collect();
+        if direct_channels.is_empty() {
+            return Err(PathFindError::PathFind(
+                "no direct channel found for source node".to_string(),
+            ));
+        }
+        // a proper hop hint for route self will limit the direct_channels to only one
+        // if there are multiple channels, we will randomly select a channel from the source node for route to self
+        // so that the following part of algorithm will always trying to find a path without cycle
+        let rand_index = rand::thread_rng().gen_range(0..direct_channels.len());
+        let (from, _, channel_info, channel_update) = direct_channels[rand_index];
```

Comment:
suggest to use `SliceRandom::choose`
```suggestion
        if let Some((from, _, channel_info, channel_update)) = direct_channels.choose(&mut thread_rng()) {

        } else {
            return Err(PathFindError::PathFind(
                "no direct channel found for source node".to_string(),
            ));
        }
```